consensus/istanbul/core/commit.go

```diff
@@ -33,28 +35,56 @@ func (c *core) sendCommit() {
 
 func (c *core) generateCommittedSeal(sub *istanbul.Subject) (blscrypto.SerializedSignature, error) {
 	seal := PrepareCommittedSeal(sub.Digest, sub.View.Round)
-	committedSeal, err := c.backend.SignBLS(seal, []byte{}, false)
+	committedSeal, err := c.backend.SignBLS(seal, []byte{}, false, false)
 	if err != nil {
 		return blscrypto.SerializedSignature{}, err
 	}
 	return committedSeal, nil
 }
 
-func (c *core) generateEpochValidatorSetData(blockNumber uint64, newValSet istanbul.ValidatorSet) ([]byte, error) {
+// Generates serialized epoch data for use in the Plumo SNARK circuit.
+// Block number and hash may be information for a pending block.
+func (c *core) generateEpochValidatorSetData(blockNumber uint64, round uint8, blockHash common.Hash, newValSet istanbul.ValidatorSet) ([]byte, []byte, bool, error) {
 	if !istanbul.IsLastBlockOfEpoch(blockNumber, c.config.Epoch) {
-		return nil, errNotLastBlockInEpoch
+		return nil, nil, false, errNotLastBlockInEpoch
 	}
+
+	// Serialize the public keys for the validators in the validator set.
 	blsPubKeys := []blscrypto.SerializedPublicKey{}
 	for _, v := range newValSet.List() {
 		blsPubKeys = append(blsPubKeys, v.BLSPublicKey())
 	}
+
 	maxNonSigners := uint32(newValSet.Size() - newValSet.MinQuorumSize())
-	epochData, err := blscrypto.EncodeEpochSnarkData(blsPubKeys, maxNonSigners, uint16(istanbul.GetEpochNumber(blockNumber, c.config.Epoch)))
-	if err != nil {
-		return nil, err
+
+	// Before the Donut fork, use the snark data encoding with epoch entropy.
+	if !c.backend.ChainConfig().IsDonut(big.NewInt(int64(blockNumber))) {
+		message, extraData, err := blscrypto.EncodeEpochSnarkData(
+			blsPubKeys, maxNonSigners,
+			uint16(istanbul.GetEpochNumber(blockNumber, c.config.Epoch)),
+		)
+		// This is before the Donut hardfork, so signify this doesn't use CIP22.
+		return message, extraData, false, err
+	}
+
+	// Retrieve the block hash for the last block of the previous epoch.
+	parentEpochBlockHash := c.backend.HashForBlock(blockNumber - c.config.Epoch)
+	if blockNumber > 0 && parentEpochBlockHash == (common.Hash{}) {
+		return nil, nil, false, errors.New("unknown block")
 	}
 
-	return epochData, nil
+	// TODO(lucas): hardcode at first, but eventually make governable
+	maxValidators := uint32(150)
```

Comment:
@lucasege wdyt?

Reply:
Sure, I can add this to a constant at the top of `core/commit.go`, unless you @mcortesi have a recommendation on an external file where this might live

Reply:
i think top of file is fine by now, with a comment of what it represents! thanks!